Use license key for downloading GeoIP db

sledgehammer999 opened the pull request:
This is a quick implementation of my suggestion in #11792

The new link serves a `.tar.gz` file instead of a `.gz` file. This archive contains 1 directory with 3 files. The db, a license text file and a copyright text file. I opted to extract and dump all of them in the `GeoIP` folder (stripping their root folder). Just in case the user goes snooping around, he will find the additional files.

It seems that the tar file format is something very simple, so I opted to implement a very-tailored to our needs parsing code instead of importing the myriad of other libraries out there. Those other libs offer a lot more than we need.
The file format is explained very well [here](https://en.wikipedia.org/wiki/Tar_(computing)#File_format). Basically it is a flat file in the form of `header+filedata+header+filedata+...`

With this change a db is loaded either by using a user provided license key or by parsing a user provided file (eg user kept a copy from their previous installation or some 3rd party has decided to distribute a copy of the db).

**Why don't we provide a copy of the db?** Well, if the contents of the db really fall under the GDPR/CCPA then I don't want to handle conforming to those 2 laws. Let each user have the legal responsibility.

PS: I wonder if MaxMind will notice the sudden influx of new user registrations if this PR goes through

sledgehammer999 commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,134 @@
+
+    // The EOF is **at least** 2 zero-ed headers. So valid data are 1 header + payload + 2 zero-ed headers
+    while ((curPos < dataSize) && (dataSize - curPos > 3 * TAR_HEADER_SIZE)
+           && !tarEOF(data.constData() + curPos) && files.size() < 11) {
```
`files.size() < 11` this is a random constraint I put here.

Chocobo1 commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,134 @@
+            return fileSize;
+        }
+
+        const char *m_data;
```
On a second thought, maybe you should use QByteArray as the building base and not a char pointer.

sledgehammer999 replied:
>maybe you should use QByteArray as the building block and not a char pointer

I don't want a unnecessary memory copies. If I had "buffer view" equivalent of std::string_view then I would use it.

glassez replied:
`QByteArray:fromRawData()` doesn't copy anything.

sledgehammer999 replied:
>QByteArray:fromRawData() doesn't copy anything.

Hmm, correct. But as I see it, the only useful "convenience" function is `QByteArray::mid()`.
Should I proceed with QByteArray?

glassez replied:
>the only useful "convenience" function is `QByteArray::mid()`.

 `QByteArray::size()` is very useful, isn't it?
>Should I proceed with QByteArray?

IMO, Yes.

sledgehammer999 replied:
on another tought, what do you say about using `libtorrent::string_view` here? Added benefit, it indicates to the caller "dude keep the buffer alive" and ensures no accidental memory copy.

`libtorrent::string_view` basically uses boost's implementation.

Chocobo1 replied:
>I don't want a unnecessary memory copies. If I had "buffer view" equivalent of std::string_view then I would use it.

We have `Utils::ByteArray::midView()`

Chocobo1 replied:
>on another tought, what do you say about using libtorrent::string_view here? 

Off-topic:
IMO string_view is more suitable for... well "strings" (ascii or utf-x variants). It is not the case here.
Here you operate on bytes which (I think) std::span in c++20 is the better matching counterpart.

sledgehammer999 replied:
>IMO string_view is more suitable for... well "strings" (ascii or utf-x variants). It is not the case here.

IMO, it's fine if you use it as generic container (no string specific methods called)

>Here you operate on bytes which (I think) std::span in c++20 is the better matching counterpart.

Then, I could use `libtorrent::span`

glassez replied:
IMO, libtorrent utils (including the types meantioned above) can't be considered as stable enough from lifetime perspective. But if `libtorrent::span` has interface similar to c++20 one then we can use it so we can easily switch to c++20 in future just by changing `lt::` to `std::`. Or we can even declare alias in our own namespace (e.g. `Utils::Span`) currently for libtorrent span and then change it to std span.

sledgehammer999 replied:
>IMO, libtorrent utils (including the types meantioned above) can't be considered as stable enough from lifetime perspective.

I don't know if I understood you correctly, but libtorrent::span is public API and is used in other public API as a type.

>But if libtorrent::span has interface similar to c++20 one then we can use it so we can easily switch to c++20 in future just by changing lt:: to std::

Exactly and it does have similar interface.

I was just ready to jump on using lt::span and then I realized we also support RC_1_1 :(
(it isn't available there)

glassez replied:
>I was just ready to jump on using lt::span and then I realized we also support RC_1_1 😞
(it isn't available there)

We can either drop RC_1_1 support or incorporate libtorrent span sources into qBittorrent sources for some time.

glassez replied:
Or just use QByteArray.

glassez replied:
>I don't know if I understood you correctly, but libtorrent::span is public API and is used in other public API as a type.

Other types can be used in further versions of libtorrent so it stops to provide its own ones.

sledgehammer999 replied:
Yeah I am leaning with QByteArray too.
Btw, there are 3rd party span implementations that are header only. This is only for FYI. I am not advocating in including them just for this.

Chocobo1 commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,134 @@
+    // The EOF is **at least** 2 zero-ed headers. So valid data are 1 header + payload + 2 zero-ed headers
+    if (data.size() < 3 * TAR_HEADER_SIZE + 1) {
+        if (ok)
+            *ok = false;
```
would be simpler if you just assigned `*ok = false;` at the top of this function.

glassez commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,134 @@
+    bool tarEOF(const char *data)
+    {
+        // static variables are auto initialized to zero
+        static char testBlock[2 * TAR_HEADER_SIZE];
```
Do you really need to store it between function calls?

sledgehammer999 replied:
See my rationale https://github.com/qbittorrent/qBittorrent/pull/11833/files#r363238761

glassez commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,134 @@
+    public:
+        TarFile(const char *data)
+            : m_data(data)
+        {}
```
Either put entire declaration into one line or split these braces too.

glassez commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,134 @@
+            return fileSize;
+        }
+
+        const char *m_data;
```
Many memory access operations use `char`, unless I'm wrong...

glassez commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,134 @@
+            return QString{m_data};
+        }
+
+        uchar type() const
```
>just write unsigned char here.

Never like these ugly c++ multiword type names.

sledgehammer999 replied:
I like the shorthand too. Plus I changed to signed int to the other place. Should I change this to (signed) char? Or should I go full (signed) int here too?

glassez replied:
I would use int.

Chocobo1 replied:
>Should I change this to (signed) char?

Seem yes, based on "typeflag field" definition on https://pubs.opengroup.org/onlinepubs/9699919799/utilities/pax.html#tag_20_92_13_06 .
Weird that they used `char` as the type there.

sledgehammer999 commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,134 @@
+    bool tarEOF(const char *data)
+    {
+        // static variables are auto initialized to zero
+        static char testBlock[2 * TAR_HEADER_SIZE];
```
The purpose was to be efficient in initializing/allocating memory.
This is what is alluded here: https://stackoverflow.com/questions/6938219/how-to-check-whether-all-bytes-in-a-memory-block-are-zero

If the above doesn't really matter, then should I do this?
```c++
char testBlock[2 * TAR_HEADER_SIZE] = {};
```

sledgehammer999 replied:
Edited previous comment.

glassez replied:
IMO, it's not so good idea to copy some memory just to check for buffer end. Why not perform integer calculations instead?

sledgehammer999 replied:
The other way I see it, is just loop in the buffer and check each byte for zero. But the SO answers say that his might not be optimal/efficient.

If there is another way, please give an example.

glassez replied:
Oops... It's `memcmp`! Forget my previous comment.

Chocobo1 replied:
>If the above doesn't really matter, then should I do this?

Yeah. 
If using QbyteArray you can use this constructor QByteArray::QByteArray(int size, char ch) and compare them with QByteArray::startsWith(const QByteArray &ba).

sledgehammer999 replied:
I think I have a better alternative to test for EOF (and possibly gracefully handle garbage archives/data):

As mentioned, a tar archive looks like this `header+filedata+header+filedata+...+empty header+empty header`
Each time we are finished with a file we are expecting either a normal header or at least 2 empty ones. And the first field on a valid header is the filename field.
So why not check if the 1st byte in the header is null, and if yes treat it as EOF?

glassez replied:
>As mentioned, a tar archive looks like this `header+filedata+header+filedata+...+empty header+empty header`

Very strange way of indicating the end of the file. It is clear that the "real" end of the file could not be used, because it was designed for streaming media, but why two empty headers and not one? And what really about not empty header but just invalid?

glassez replied:
>So why not check if the 1st byte in the header is null, and if yes treat it as EOF?

At least we can use simplified algorithm.
1. If next byte is not null try to load header and so on...
2. If next byte is null stop processing (but check whether there's valid EOF to distinguish between correct/broken files).

sledgehammer999 replied:
>If next byte is not null try to load header and so on...

Even if next byte is not null, we can't really extract a filename from that. There isn't supposed to be padding* between filedata and the next header.

*padding exists at the end of filedata to pad to full 512bytes block.

>but check whether there's valid EOF to distinguish between correct/broken files

Hmm, this check needs to done inside `TarFile::fileData()`. But a variation of it. Not detect null bytes but actually test that there is enough space left in the buffer for the filesize indicated by the header.

glassez replied:
>Even if next byte is not null, we can't really extract a filename from that. 

I mean first byte of (possible) next header.
E.g.:
1. Try to load header.
2. If there isn't enough data - error.
3. Perform the "optimize" check for null in first byte of header.
4. If it is null, check for correct EOF at this position. If there's no valid EOF - error.
5. If it isn't null try to parse header and perform other checks on it. If header is invalid - error.
6. Try to load file.
7. If there isn't enough data - error.
8. Perform some checks on file data (e.g. checksum comparison). If it fails - error.
9. Try to shift to next header (i.e. skip padding). If there's no enough data - error.
10. Start from beginning.

Have I missed something? I didn't read full format specs.

sledgehammer999 commented:
On another note: The tar header includes a `checksum` field. Should I validate the checksum too?

sledgehammer999 commented:
>Have I missed something? I didn't read full format specs.

I got the basic idea. I'll see if I can push something later today.

sledgehammer999 commented:
I have pushed a new version. I almost rewrote the whole `tar.cpp`.

sledgehammer999 commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,230 @@
+                continue;
+
+            const int filePos = curPos - fileSize - padding;
+            files.insert(file.name(), {data.constData() + filePos, fileSize});
```
Should I construct the QByteArray here using fromRawData() for extra efficiency?

glassez replied:
You can't (shouldn't) do it! Then returned data is valid only until input data isn't disposed.

sledgehammer999 commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,230 @@
+                continue;
+
+            const int filePos = curPos - fileSize - padding;
+            files.insert(file.name(), {data.constData() + filePos, fileSize});
```
Note to self: Must check that there are `fileSize` bytes in the buffer.

glassez commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,230 @@
+    };
+}
+
+
```
Redundant empty line.

glassez commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,230 @@
+                break;
+
+            if (!file.isValid()) {
+                LogMsg(QCoreApplication::translate("Utils::Tar",
```
It's wrong. Here you just unpacking some tar archive (it's abstract utility function). You should perform any logging at the application level.

glassez commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,230 @@
+    int dataSize = data.size();
+    QHash<QString, QByteArray> files;
+
+    while ((curPos < dataSize) && files.size() < 11) {
```
It's weird to see such limitation in abstract utility function... does tar format limit it in this way?

LordNyriox replied:
@glassez:  According to @sledgehammer999's comment on an earlier version of this code: 

> `files.size() < 11` this is a random constraint I put here. 

**Source:**  [<https://github.com/qbittorrent/qBittorrent/pull/11833#discussion_r363149053>]

glassez replied:
@LordNyriox, sorry, what do you want to say me? I saw that comment.

LordNyriox replied:
@glassez: 

> what do you want to say me? 

What I mean is that sledgehammer described it as a "random" constraint. 

In this context, "random" most likely means "chosen at random" (in other words, a decision made *without* any explicit reason for it).

glassez replied:
I'm sorry, I can't understand what you're trying to explain to me...
I understand the meaning of word "random" (and even what it means *in this context*), as well as what and why @sledgehammer999 wants to do here. I'm just pointing out the fact that it was done in the wrong place (at the wrong level of logic). That's all.

glassez replied:
Also please read https://github.com/qbittorrent/qBittorrent/pull/11833#issuecomment-571931231.

sledgehammer999 replied:
I removed this "security" constraint in the new code.

glassez commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,230 @@
+        // undefined behavior.
+        int customStrlen(const QByteArray &str) const
+        {
+            int sz = 0;
```
```c++
const int len = str.indexOf(0);
return (len >= 0) ? len : str.length();
```

glassez commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,230 @@
+        // the whole buffer when we already know the upper limit. Also the buffer
+        // isn't guaranteed to contain a null byte and std::strlen() will have
+        // undefined behavior.
+        int customStrlen(const QByteArray &str) const
```
Maybe declare it as `Utils::ByteArray::strlen()`?
Anyway I dislike "custom" prefix here.

glassez commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,230 @@
+}
+
+
+QHash<QString, QByteArray> Utils::Tar::decompress(const QByteArray &data, bool *ok)
```
It's better to use exceptions instead of "ok". Then you can provide more error info to the caller side.

sledgehammer999 replied:
I took another approach. Have a look.

glassez commented:
General review comment.
I would implement it in form of "stream reader". This is a fairly effective pattern in terms of efficiency. In addition, it is better suited to our needs - we can easily apply the necessary restrictions without breaking the abstract unpacking mechanism itself. E.g. (pseudo code):
```c++
TarReader reader {data};
while (reader.readNext()) {
    const TarEntry entry = reader.entry();
    // here TarEntry struct contains parsed file data and metadata (i.e. name)
    // so we can do something with it
}

// here we have reached either end of file or error
if (reader.error()) {
    // do something...
}
```

glassez commented on src/base/net/geoipmanager.cpp:
```diff
@@ -110,13 +111,20 @@ void GeoIPManager::loadDatabase()
 {
+    if (m_dbCompleteUrl.isEmpty()) {
+        if (!m_geoIPDatabase)
+            Logger::instance()->addMessage(tr("No GeoIP license key has been configured nor a GeoIP database file has been found."), Log::WARNING);
```
IMO, log message in wrong place (under wrong condition). When user doesn't provide license key we shouldn't even trigger database update.
The only valid error case is when user enables "resolve peer country" feature but there are no license key or db file provided.

glassez replied:
How about this one?

sledgehammer999 replied:
Hmm, I am not sure I agree with you. `manageDatabaseUpdate()` "manages" the db update. Should calling code do the error checking for it? My position is that the error checking must be done by `manageDatabaseUpdate()` itself.

glassez replied:
My comment is more about "what we should consider as an error" than "what place in the code we should handle this error".

glassez replied:
Possible algorithm (when "resolve peer country" is enabled):
1. If license key is set by user try to update database and log errors if they are. Otherwise just skip database update step.
2. If database isn't loaded by previous step try to load it from file and log errors if they are.

sledgehammer999 replied:
Now I understand your objection better. I pushed an update and hopefully manages things more fully and clearer.
PS: The logic is to first load any existing db and then perform db update if possible and necessary.

glassez commented on src/base/net/geoipmanager.cpp:
```diff
@@ -392,10 +400,16 @@ QString GeoIPManager::CountryName(const QString &countryISOCode)
+    const bool enabled = pref->resolvePeerCountries();
+    QString license = pref->geoIPLicenseKey();
+    QString dbCompleteUrl = license.isEmpty() ? QString{} : QString{DATABASE_URL + license};
+    if (m_enabled != enabled
```
Please enclose subexpressions.

glassez replied:
```c++
if ((m_enabled == enabled) && (m_dbCompleteUrl == dbCompleteUrl))
     return;
```

sledgehammer999 commented on src/base/utils/tar.h:
```diff
@@ -0,0 +1,42 @@
+{
+    namespace Tar
+    {
+        QHash<QString, QByteArray> decompress(const QByteArray &data, bool *ok = nullptr);
```
renamed to `untar()`

sledgehammer999 commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,230 @@
+
+        enum Types : unsigned char
+        {
+            REG_FILE_NULL = '\0',
```
Other functions require passing the `'\0'` literal because of the compiler. It doesn't know which overload to call (eg for startsWith()).
Furthermore, in the spec these are treated as ASCII values, so for keeping things uniform, I will leave it as ASCII NUL.

sledgehammer999 commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,230 @@
+
+            // Invalid header: The first header field represents the filename.
+            // If the 1st byte is null then we can't extract a name. So the whole header becomes invalid to us.
+            m_valid = m_header.at(0) != '\0';
```
compiler considers it an ambiguous call.

sledgehammer999 commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,230 @@
+        }
+
+        bool isNull() const { return m_null; }
+        bool isValid() const { return m_valid; }
```
I took another approach. Please take a look.
We need to differentiate between fully Null entry and an malformed one.
If you can come up with another way, let me hear it.

sledgehammer999 commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,230 @@
+                calculatedChecksum += static_cast<uchar>(m_header.at(i));
+
+            const int checkSum = octalToDec(Utils::ByteArray::midView(m_header, TAR_CHECKSUM_OFFSET, TAR_CHECKSUM_SIZE));
+            m_null = (calculatedChecksum == TAR_CHECKSUM_SIZE * ' ') && (checkSum == 0);
```
Something that you might not know: When checksum is calculated the bytes belonging to the `checksum` header are considered to have each the value 32 (ASCII space).
So in a fully null header the `calculatedChecksum` variable will have the value (8 * 32) AND the stored checksum in the checksum field (from `checkSum` variable) will have the value 0.

sledgehammer999 commented:
Pushed new version. Take a another look. It is supposed to work as a stream reader. Not sure if I got the interface right.
Please focus revies on `tar.cpp`. Once the code is finalized, we can focus on the other files.

Chocobo1 commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,298 @@
+
+            const int fileSize =  m_entry.size();
+            const int filePos = m_pos + TAR_HEADER_SIZE;
+            return {m_data.constData() + filePos, fileSize};
```
This is copying the data, you should avoid it.
Generally speaking, no-copy is the key for building fast applications and a requirement for modern apps IMO.

glassez replied:
>no-copy is the key for building fast applications and a requirement for modern apps IMO.

It also provides many opportunities for errors to occur.
It is especially dangerous to return such "references" to the internal data of the class (this is generally unacceptable and is a design error, IMO). One exception is when class instance operates on some outer data (so caller side owns this data and cares about its lifetime and lifetime of "references" into it).
So if you really want to implement "no-copy" optimization you need to think a hundred times about all the details of the implementation, in order to protect the caller as much as possible.

Chocobo1 replied:
>It also provides many opportunities for errors to occur.

Yes and I would hope/expect there will be nicely designed class interface to avoid that.

Offtopic I think is worth sharing, here is a library (in rust lang) that implements a tar reader: https://github.com/alexcrichton/tar-rs#reading-an-archive (ignore the `unwrap()` for easier reading)
You can see a `File` object is passed to `tar::Archive` constructor and thus (normally) the tar reader won't outlive the `File` lifetime.

(more offtopic) And another nice thing that library has:
From https://docs.rs/tar
>Great strides are taken to ensure that an archive is never required to be fully resident in memory, and all objects provide largely a streaming interface to read bytes from.

sledgehammer999 replied:
>One exception is when class instance operates on some outer data (so caller side owns this data and cares about its lifetime and lifetime of "references" into it).

In case it wasn't clear: The buffer with the tarred data is given by the caller of `Utils::Tar::untar()`.

glassez replied:
>Yes and I would hope/expect there will be nicely designed class interface to avoid that.

There is no way to avoid that. You can only warn about the danger (in code comments) and hope that caller can correctly handle it.

glassez replied:
>You can see a File object is passed to tar::Archive constructor and thus (normally) the tar reader won't outlive the File lifetime.

File stream reading is different thing so let's not consider it here.

sledgehammer999 replied:
I may be wrong, but in addition to leaving a code comment we could do this:
```c++
QHash<QString, QByteArray> untar(const QByteArray &&, QString &) = delete;
```

This will ensure that the caller will have a non-temporary buffer. It will fail only if he does weird things in the calling site and its buffer goes out of scope before the return value. eg something like this:
```c++
void callingSite()
{
    QHash<QString, QByteArray> files;

    {
        QString error;
        const QByteArray tarred = getTarredData();
        files = Utils::Tar::untar(tarred, error);
    }

    write_to_disk(files.at(0)); // undefined behavior
}
```

Chocobo1 replied:
>File stream reading is different thing so let's not consider it here.

That is just one instance of the `Read` trait. IMO the library is still valuable as a reference when designing class interface (more or less) as the essence isn't all that different.

>>Yes and I would hope/expect there will be nicely designed class interface to avoid that.
>
>There is no way to avoid that. 

There is always ways to shoot oneself in the foot in c++ but a good interface will make it hard to do so.

>I may be wrong, but in addition to leaving a code comment we could do this:

Yeah, that will help a bit. Although there are still easy ways to go wrong:
```c++
QByteArray *tarred = new QByteArray();
getTarredData(tarred);

QString error;
QHash<QString, QByteArray> files = Utils::Tar::untar(*tarred, error);

delete tarred;
write_to_disk(files.at(0));  // undefined behavior
```

glassez replied:
Your examples are too deliberate to make a mistake in this way (if someone has read the appropriate warning).
There is a more plausible situation when you save the obtained data somewhere (especially implicitly, e.g. create some dynamically allocated object and pass it there via constructor or some setter) and then the function that owns the source buffer returns.

glassez replied:
The problem is there are no differences at the class interface between QByteArray that owns the data and that doesn't. So if some class need accepts QByteArray to store it internally there is no way to prevent error case from this class.
IMO, the better design is when we have different classes for data and references to its parts (e.g. QString and QStringRef). So if caller side has QStringRef and calling side accepts QString we should convert it explicitly (making deep copy of referenced data).
Unfortunately we can't use QString for raw data.

Chocobo1 commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,298 @@
+        if (!(reader.type() == TarEntry::REG_FILE_ASCII || reader.type() == TarEntry::REG_FILE_NULL))
+            continue;
+
+        files.insert(reader.name(), reader.fileData());
```
I know this was discussed before. However I would still like the returned values to be QByteArray views.
Adding a comment about this function is returning a view (that the input `data` is required to be alive in client code in order for the returned value to be meaningful) would suffice for me. And then the return type will be expected to become `QHash<QString, const QByteArray>`

Chocobo1 commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,298 @@
+        int size() const { return isValid() ? octalToDec(Utils::ByteArray::midView(m_header, TAR_FILESIZE_OFFSET, TAR_FILESIZE_SIZE)) : 0; }
+
+    private:
+        int octalToDec(const QByteArray &data) const
```
This can be extracted to anonymous namespace.
And `int strlen(const QByteArray &str)` too.

Chocobo1 commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,298 @@
+        // the whole buffer when we already know the upper limit. Also the buffer
+        // isn't guaranteed to contain a null byte and std::strlen() will have
+        // undefined behavior.
+        int strlen(const QByteArray &str) const
```
IMO better differentiate the name with std::strlen() to avoid confusion.

glassez replied:
There is `uint qstrnlen(const char *str, uint maxlen)` that you can use instead (or at least as underlying one).

Chocobo1 commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,298 @@
+
+        bool isNull() const { return m_null; }
+        bool isValid() const { return m_valid; }
+        bool isVerified() const { return m_verified; }
```
Regarding `m_null`, `m_valid`, `m_verified`. Does these status intersect? I mean is it possible to have any two of them to be true at the same time?
If not it is better to tie them together into a separate type.

sledgehammer999 replied:
Maybe I could merge `m_valid` and `m_verified`. Now `m_valid == true` only when the first byte isn't null. And `m_verified == true` when the checksum is correct. This allows to have more specific error output. But I could merge those 2 so `m_valid == true` only when 1st byte != null AND checksum is correct.

`m_null == true` when all bytes are null. It is used a special marker for EOF. I don't think I can combine it with the other 2 bools.

glassez replied:
Header can be either Invalid, Null or Regular so you just need enum with 3 values.

glassez commented on src/base/utils/tar.h:
```diff
@@ -0,0 +1,42 @@
+{
+    namespace Tar
+    {
+        QHash<QString, QByteArray> untar(const QByteArray &data, QString &errorString);
```
I would make TarStreamReader public instead of (or in addition to) this wrapper.

glassez commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,298 @@
+        bool isValid() const { return m_valid; }
+        bool isVerified() const { return m_verified; }
+        Types type() const { return isValid() ? static_cast<Types>(m_header.at(TAR_FILETYPE_OFFSET)) : INVALID; }
+        int size() const { return isValid() ? octalToDec(Utils::ByteArray::midView(m_header, TAR_FILESIZE_OFFSET, TAR_FILESIZE_SIZE)) : 0; }
```
These two method names are ambiguous. They should be "fileType" and "fileSize".

glassez commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,298 @@
+    const int TAR_CHECKSUM_SIZE = 8;
+    const int TAR_FILENAME_PREFIX_SIZE = 155;
+
+    class TarEntry
```
This class has inconsistent internal behavior. Why do you parse header partially when create instance and then parse the rest on demand? I see no point in this.

sledgehammer999 replied:
>Why do you parse header partially when create instance and then parse the rest on demand?

What do I do partially? You mean the chechsum check and null check?

If I parse the rest on ctor, that means I would have to have extra member variables to hold a copy of the data.

glassez replied:
>If I parse the rest on ctor, that means I would have to have extra member variables to hold a copy of the data.

What's the problem with it? If header is invalid (or null) they will be empty. If header is valid you still copy this data when access it from caller side (since it's pointless to create this object and then never access its fields).

sledgehammer999 replied:
>What's the problem with it?

Increased class size.
I want to stay with the parse-on-demand approach and not have unnecessary copies of data laying around. Something akin to `bedecode_token/bedecode_node` (somewhat inspired by [this](https://blog.libtorrent.org/2015/03/bdecode-parsers/))

Usually the user of the class will call only once each method and store the result. So why make the class also store a copy?

glassez replied:
>Usually the user of the class will call only once each method and store the result.

No problem (unless you need to call some of them twice).
But in this case the class itself becomes redundant (now it is like a "thing in itself" that only takes care of its own invariant in each method). You can easily apply any parsing function on header data separately.

glassez commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,298 @@
+        {
+            REG_FILE_NULL = '\0',
+            REG_FILE_ASCII = '0',
+            // We don't care about the other file types
```
I would still insist on making this parser as more generic as possible (as long as it does not require excessive effort). Just return "type" as is (maybe just expanding it to `int` as discussed earlier). Let upper level decide whether it should care about some file type or not.

glassez replied:
This one is neither satisfied nor even answered.

glassez replied:
>I would still insist on making this parser as more generic as possible 

If you can't make a generic tar parser for some reason, just move it in "geoipmanager.cpp" and make it a "private" parser that can parse only particular subset of .tar files corresponding to geoip database distro.

glassez commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,298 @@
+        {
+            // only the ASCII characters from '0' to '7' are valid
+            int i = 0;
+            for (; i < data.size(); ++i) { // skip leading invalid bytes
```
Looks suspicious. IMO, you provide some kind of undefined behaviour.
You should fail on really invalid characters.
If spaces at beginning are acceptable (i.e. "   354" is valid octal number) you should skip spaces explicitly. But, e.g., "gdi754" is invalid octal number and it should cause an exception.

sledgehammer999 replied:
Sorry for the lack of activity. I'll probably push something the next days based on the various comments. On this particular one:
IIRC the specs calls that the field is an octal number padded with leading zeroes (all ASCII chars). So I am going to interpret it this ways:
```c++
int octalToDec(const QByteArray &data) const
{
int fileSize = 0;

  for (int i=0; i < data.size(); ++i) {
    char byte = data.at(i);
      if (byte < '0' || byte > '7') // skip trailing invalid bytes
        break;

    fileSize = (8 * fileSize) + (byte - '0');
  }

return fileSize;
}
```

glassez replied:
>IIRC the specs calls that the field is an octal number padded with leading zeroes (all ASCII chars).

If it's true, then your current algorithm is correct, but it can be used only internally by tar parser (so we should never make it shareable). Or you can split it into two functions: first (can be shareable util) parses valid octal notation (and breaks at any invalid character), second (private) skips leading invalid characters and apply the first one on the rest.

sledgehammer999 commented:
I think I addressed the pending comments.

**About returning a view:** @glassez is right. I was intending to introduce an additional function named `untarView()`, hoping that the naming will help future reviewers to catch improper use. While I was going to wire it up in `geoipmanager` class I realized that the DB buffer is further passed down to other classes (eg the db parser/decoder).
IMO, it is incredibly easy in the future to change something in these "downstream" classes that will assume that we own the buffer or forget to copy it when assigning. A proper type(eg string_view/span) is IMO obligatory for safe usage.
So I opted to always return a copy of the data for the time being.

glassez commented on src/base/utils/tar.h:
```diff
@@ -0,0 +1,112 @@
+    {
+        QHash<QString, QByteArray> untar(const QByteArray &data, QString &errorString);
+
+        class Header
```
Seems this class is private helper. Please move it out of here.

glassez commented on src/base/utils/tar.h:
```diff
@@ -0,0 +1,112 @@
+            explicit StreamReader(const QByteArray &data);
+            explicit StreamReader(const QByteArray &&data) = delete;
+
+            Header::Types readNext();
```
Why don't you go the common way? Just return the status of successful reading as a bool. Other information is meaningless at this point.

glassez commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,240 @@
+#include <QCoreApplication>
+#include <QHash>
+
+#include "base/logger.h"
```
Leftover

NotTsunami commented on src/base/net/geoipmanager.cpp:
```diff
@@ -420,7 +433,16 @@ void GeoIPManager::downloadFinished(const DownloadResult &result)
+
     if (geoIPDatabase) {
         if (!m_geoIPDatabase || (geoIPDatabase->buildEpoch() > m_geoIPDatabase->buildEpoch())) {
             if (m_geoIPDatabase)
```
Slightly unrelated to this PR but there are instances in geoipmanager where you can omit the null check here, [delete in c++ is a noop on null pointers](https://isocpp.org/wiki/faq/freestore-mgmt#delete-handles-null). If you want to keep that seperate I'm more than happy to address that in a seperate PR.

NotTsunami replied:
Addressed in #11961

sledgehammer999 commented:
@glassez I addressed your comments. The only thing I am not sure is about class `Header`. I made its implementation private. But if you meant to take it out of the header file entirely, then I would have to implement `StreamReader::m_entry` as a pointer and constantly new/delete it in the .cpp. Is that what you meant? Or any other way?

glassez commented on src/base/net/geoipmanager.cpp:
```diff
@@ -392,10 +400,15 @@ QString GeoIPManager::CountryName(const QString &countryISOCode)
+    const Preferences *const pref = Preferences::instance();
+    const bool enabled = pref->resolvePeerCountries();
+    QString license = pref->geoIPLicenseKey();
+    QString dbCompleteUrl = license.isEmpty() ? QString{} : QString{DATABASE_URL + license};
```
Add space between type name and `{`.

glassez commented on src/base/net/geoipmanager.cpp:
```diff
@@ -420,7 +433,16 @@ void GeoIPManager::downloadFinished(const DownloadResult &result)
+        if (!iter.key().endsWith(GEOIP_FILENAME))
+            continue;
+
+        geoIPDatabase = GeoIPDatabase::load(iter.value(), error);
```
Shouldn't you break the loop after you've found database file?

glassez commented on src/base/preferences.cpp:
```diff
@@ -896,6 +896,16 @@ void Preferences::resolvePeerCountries(const bool resolve)
 
+QString Preferences::geoIPLicenseKey() const
+{
+    return value("Preferences/Connection/GeoIPLicenseKey").toString();
```
I wouldn't use legacy configuration structure for new settings (it was bound to the Options dialog layout). IMO, it's better to have something like "GeoIP/LicenseKey".

glassez commented:
>The only thing I am not sure is about class Header. I made its implementation private. But if you meant to take it out of the header file entirely, then I would have to implement StreamReader::m_entry as a pointer and constantly new/delete it in the .cpp. Is that what you meant? Or any other way?

I would have some pure struct here to store current entry fields (or even just appropriate fields in StreamReader itself) and keep all the logic in StreamReader class. Otherwise you can have some private parser in .cpp (if you really need to separate parsing logic) that parses the entry header data and fills the appropriate fields of StreamReader (e.g. returns the struct with parsed fields).
You don't really need to store parser instance in StreamReader class. It can be created just before parsing and destroyed after you store its result. Otherwise you pollute the header with too many implementation details.

sledgehammer999 commented on src/base/utils/tar.h:
```diff
@@ -0,0 +1,113 @@
+        class StreamReader
+        {
+        public:
+            enum class Type : unsigned char
```
The type of the header can point to other things than files (eg symlink, directory, etc)

glassez replied:
But it's still not "type of Stream Reader"...
So maybe "EntryType"?

sledgehammer999 commented on src/base/utils/tar.h:
```diff
@@ -0,0 +1,113 @@
+            enum class Type : unsigned char
+            {
+                RegFileNull = '\0',
+                RegFileAscii = '0',
```
Regular file. This can be denoted by 2 separate values. Either null or ascii zero.
Should I write it out fully? eg `RegularFileNull`

glassez replied:
Are there any other (non regular) files so you can't omit "regular" keyword? If there's one please use "Regular" instead of "Reg".

sledgehammer999 commented on src/base/utils/tar.h:
```diff
@@ -0,0 +1,113 @@
+            QByteArray fileData() const;
+            Type fileType() const;
+
+            bool atEnd() const { return m_end; }
```
And IIRC you were the one that advocated that such simple instructions can be implemented in headers. And I think there is actually used in some headers now. Am I wrong?

glassez replied:
>you were the one that advocated that such simple instructions can be implemented in headers.

Yes. And I still didn't change my mind. But I don't use it since it's forbidden by qBittorrent Coding Guidelines. Maybe it's time to reconsider it?

sledgehammer999 replied:
Sorry. I had the wrong idea in my head. Apparently we only do member variable initializations in the header. I don't know if this is constrained to only structs and very simple classes.

>Maybe it's time to reconsider it?

I don't know. I don't even remember my exact reasoning for being against this. The only thought I have for being against it now: Consistency in a class. When you're looking at a .cpp you probably want to see all member function implementations there and not go hunting them down in the header. I admit this isn't a strong case, and personally I don't have a strong position on this.

Chocobo1 replied:
>Consistency in a class. When you're looking at a .cpp you probably want to see all member function implementations there and not go hunting them down in the header. 

Also it is easier to read the function signatures (in header) without the function bodies.

glassez replied:
Okay, whatever you want... I'm used to it.

sledgehammer999 commented:
>I would have some pure struct here to store current entry fields (or even just appropriate fields in StreamReader itself) and keep all the logic in StreamReader class. Otherwise you can have some private parser in .cpp (if you really need to separate parsing logic) that parses the entry header data and fills the appropriate fields of StreamReader (e.g. returns the struct with parsed fields).
You don't really need to store parser instance in StreamReader class. It can be created just before parsing and destroyed after you store its result. Otherwise you pollute the header with too many implementation details.

Yeah, after so many iterations it doesn't make to sense to keep the helper class for the current implementation. I removed the class and implemented it via anonymous namespace helper functions and StreamReader member variables.

an0n666 commented:
So what happens after this gets merged?
Will latest Qbit installer ship with the latest GeoLite db as fallback?
Or are you making going to make everyone register at maxmind for a license key (lol)
I don't see how this will be useful for majority of the users who would probably never figure out how to do it.
I think qbit should include the database in the installer. Downloading using individual license will probably even flood Maxmind with new registrations which are completely unnecessary.

sledgehammer999 commented:
@an0n666 

>With this change a db is loaded either by using a user provided license key or by parsing a user provided file (eg user kept a copy from their previous installation or some 3rd party has decided to distribute a copy of the db).
>
>**Why don't we provide a copy of the db?** Well, if the contents of the db really fall under the GDPR/CCPA then I don't want to handle conforming to those 2 laws. Let each user have the legal responsibility.

glassez commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,248 @@
+        return false;
+
+    int dataSize = m_data.size();
+    if (m_pos >= dataSize)
```
When readNext returns false it should be either error or end-of-file case, isn't it?

sledgehammer999 replied:
Your comment helped me realize another solution. I dropped the `bool m_end` member variable. When the parser encounters the empty header (marker for EOF) it sets `m_pos` to `m_data.size()`. And `atEnd()` tests for that.

glassez commented on src/base/net/geoipmanager.cpp:
```diff
@@ -392,10 +401,15 @@ QString GeoIPManager::CountryName(const QString &countryISOCode)
         m_enabled = enabled;
-        if (m_enabled && !m_geoIPDatabase) {
+        if (m_enabled
+            && (!m_geoIPDatabase || (m_dbCompleteUrl != dbCompleteUrl))) {
```
These conditions seem to be redundant.
Due to conditions in "if" above it was either disabled before (so m_geoIPDatabase was nullptr) or it had different database url.

sledgehammer999 replied:
>either disabled before (**so m_geoIPDatabase was nullptr**)

I didn't get into the specifics because the deleted check included `!m_geoIPDatabase` already. Of the top of my head I can't find a flow where it results that it was disabled but with non-nullptr pointer (m_geoIPDatabase). Maybe it was put there as an extra precaution?

Should I try to remove it? (and from the else if below it?)

glassez replied:
>Should I try to remove it? (and from the else if below it?)

It can be written as:
```c++
if (m_enabled) {
}
else {
}
```
without any other conditions.
>Maybe it was put there as an extra precaution?

Maybe...
Anyway we shouldn't handle it as regular case (unless we really want to keep database loaded after feature is disabled).

glassez commented:
>The logic is to first load any existing db and then perform db update if possible and necessary.

IIRC, it's existing logic, although it seems weird to me currently...

an0n666 commented:
Have you looked into alternatives like db-ip? 
https://db-ip.com/db/lite.php
@sledgehammer999

sledgehammer999 commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,248 @@
+    const int padding = (m_fileSize > 0) ? (TAR_HEADER_SIZE - (m_fileSize % TAR_HEADER_SIZE))
+                                       : 0;
+
+    if (m_hasParsingStarted)
```
I know.
It was either this `bool` or an `int` to hold to start position of the filedata. I chose bool in an effort to minimize class size in memory. But now I see that my last bool simply introduces padding at the end of the class. So instead of padding I'll use the `int` which should fit nicely there.

glassez replied:
> I chose bool in an effort to minimize class size in memory. 

It's invalid reason to make logic inconsistent.

sledgehammer999 commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,248 @@
+    }
+
+    const int padding = (m_fileSize > 0) ? (TAR_HEADER_SIZE - (m_fileSize % TAR_HEADER_SIZE))
+                                       : 0;
```
How so? Do you mean the formatting or what the code does?

glassez replied:
>Do you mean the formatting 

Yes.

sledgehammer999 commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,248 @@
+        // If the 1st byte is null then we can't extract a name. So the whole header becomes invalid to us.
+        if ((state == State::Valid) && header.startsWith('\0'))
+            return State::Invalid;
+        else if (state != State::Valid)
```
I would like to keep it this way. If I use a naked `if` then I'll have to introduce an empty line between the 2 tests for readability.

glassez replied:
> I'll have to introduce an empty line between the 2 tests for readability.

Is it a problem?

sledgehammer999 commented on src/base/utils/tar.cpp:
```diff
@@ -0,0 +1,248 @@
+        if ((calculatedChecksum == TAR_CHECKSUM_SIZE * ' ')
+            && (checkSum == 0))
+            return State::Empty;
+        else if (calculatedChecksum == checkSum)
```
same

sledgehammer999 commented:
>Have you looked into alternatives like db-ip?

Yes, but I don't think I came across that one. It seems to serve the db in the same format as MaxMind.

@glassez @Chocobo1 it seems that all we have to do is change the URL. Should this PR be dropped?
Do we have any reason on not using that service instead?

Chocobo1 commented:
>Do we have any reason on not using that service instead?

If it works, then why not?

Also this https://db-ip.com/db/download/ip-to-country-lite
>The free IP to Country Lite database by DB-IP is licensed under a Creative Commons Attribution 4.0 International License.
>
>You are free to use this IP to Country Lite database in your application, provided you give attribution to DB-IP.com for the data.

glassez commented:
>>Do we have any reason on not using that service instead?
>If it works, then why not?

:+1:

an0n666 commented:
> > Have you looked into alternatives like db-ip?
> 
> Yes, but I don't think I came across that one. It seems to serve the db in the same format as MaxMind.
> 
> @glassez @Chocobo1 it seems that all we have to do is change the URL. Should this PR be dropped?
> Do we have any reason on not using that service instead?

The URL, DB_TYPE[], METADATA_BEGIN_MARK[] and maybe some other things as well.

sledgehammer999 commented:
>Apparently they don't use a static URL like maxmind. It changes according to month.

Damn. Actually. I have implemented all necessary changes locally. I was waiting to polish the docs a bit. I did it in a hurry and didn't realize that URL has the date encoded in it. Damn it.
If I find time, I may send them an e-mail about this.
Alternatively, we could try our luck by "guessing" the next month's URL.

an0n666 commented:
Guessing might work. 
dbip-country-lite-2020-%month.mmdb.gz

an0n666 commented:
You could also just use one subdomain from qbitorrent.org to redirect to latest version every month.

sledgehammer999 commented:
Alternative PR implemented: #12014

FranciscoPombal commented:
> Alternative PR implemented: #12014

So is this superseded or are there any alternative plans for this?

sledgehammer999 commented:
>So is this superseded or are there any alternative plans for this?

Superseded and will be closed once the other one is merged.

sledgehammer999 commented:
Closing as superseded by #12014
